src/workerd/api/form-data.c++

```diff
@@ -157,109 +235,132 @@ void parseFormData(kj::Maybe<jsg::Lock&> js,
-        // the isolate's external memory.
-        auto bytes = kj::heapArray<kj::byte>(message.asBytes());
-        data.add(FormData::Entry{kj::mv(name),
-          jsg::alloc<File>(kj::mv(bytes), KJ_ASSERT_NONNULL(kj::mv(filename)),
-              kj::str(type.orDefault(nullptr)), dateNow())});
-      }
-    }
+    callback(name, filename.map([](auto& str) { return str.asPtr(); }), type, message.asBytes());
   }
 }
 
-kj::OneOf<jsg::Ref<File>, kj::String> blobToFile(jsg::Lock& js,
-    kj::StringPtr name,
-    kj::OneOf<jsg::Ref<File>, jsg::Ref<Blob>, kj::String> value,
-    jsg::Optional<kj::String> filename) {
-  auto fromBlob = [&](jsg::Ref<Blob> blob) {
-    kj::String fn;
-    KJ_IF_SOME(f, filename) {
-      fn = kj::mv(f);
-    } else {
-      fn = kj::str(name);
-    }
-    // The file is created with the same data as the blob (essentially as just
-    // a view of the same blob) to avoid copying the data.
-    return jsg::alloc<File>(
-        blob.addRef(), blob->getData(), kj::mv(fn), kj::str(blob->getType()), dateNow());
-  };
+kj::Array<FormData::EntryWithoutLock> FormData::parseWithoutLock(
+    kj::ArrayPtr<const char> rawText, kj::StringPtr contentType) {
+  kj::Vector<FormData::EntryWithoutLock> data;
 
-  KJ_SWITCH_ONEOF(value) {
-    KJ_CASE_ONEOF(file, jsg::Ref<File>) {
-      if (filename == kj::none) {
-        return kj::mv(file);
-      } else {
-        // Need to substitute filename.
-        return fromBlob(kj::mv(file));
+  KJ_IF_SOME(parsed, MimeType::tryParse(contentType)) {
+    auto& params = parsed.params();
+    if (MimeType::FORM_DATA == parsed) {
+      auto& boundary = JSG_REQUIRE_NONNULL(params.find("boundary"_kj), TypeError,
+          "No boundary string in Content-Type header. The multipart/form-data MIME "
+          "type requires a boundary parameter, e.g. 'Content-Type: multipart/form-data; "
+          "boundary=\"abcd\"'. See RFC 7578, section 4.");
+
+      parseFormDataImpl(rawText, boundary,
+          [&](kj::StringPtr name, kj::Maybe<kj::StringPtr> maybeFilename,
+              kj::Maybe<kj::StringPtr> maybeType, kj::ArrayPtr<const kj::byte> message) mutable {
+        KJ_IF_SOME(filename, maybeFilename) {
+          data.add(FormData::EntryWithoutLock{
+            .name = kj::str(name),
+            .filename = kj::str(filename),
+            .type = maybeType.map([](kj::StringPtr str) { return kj::str(str); }),
+            .value = kj::heapArray<kj::byte>(message),
+          });
+        } else {
+          data.add(FormData::EntryWithoutLock{
+            .name = kj::str(name),
+            .value = kj::str(message),
+          });
+        }
+      });
+      return data.releaseAsArray();
+    } else if (MimeType::FORM_URLENCODED == parsed) {
+      // Let's read the charset so we can barf if the body isn't UTF-8.
+      //
+      // TODO(conform): Transcode to UTF-8, like the spec tells us to.
+      assertUtf8(params);
```

Comment:
This assertion seems like a breaking change, isn't it?

Reply:
Nope, I just encapsulated the existing check into a reusable function